src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1065 @@
+ShardArrayToList(ShardInterval **shardArray, int length)
+{
+	List *shardIntervalList = NIL;
+	int shardIndex;
+
+	for (shardIndex = 0; shardIndex < length; shardIndex++)
+	{
+		ShardInterval *shardInterval =
+			shardArray[shardIndex];
+		shardIntervalList = lappend(shardIntervalList, shardInterval);
+	}
+
+	return shardIntervalList;
+}
+
+
+/*
+ * Perform pruning for a single PruningInstance.
+ */
+static List *
+PruneOne(DistTableCacheEntry *cacheEntry, ClauseWalkerContext *context,
+		 PruningInstance *prune)
+{
+	ShardInterval *shardInterval = NULL;
+
+	if (prune->alwaysFalse)
+	{
+		return NIL;
+	}
+
+	/*
+	 * For an equal constraints, if there's no overlapping shards (always the
+	 * case for hash and range partitioning, sometimes for append), can
+	 * perform binary search for the right interval.  That's usually the
+	 * fastest, so try that first.
+	 */
+	if (prune->equalConsts &&
+		!cacheEntry->hasOverlappingShardInterval)
+	{
+		shardInterval = FastShardPruning(cacheEntry, prune->equalConsts->constvalue);
+
+		/*
+		 * If pruned down to nothing, we're done. Otherwise see if other
+		 * methods prune down further / to nothing.
+		 */
+		if (!shardInterval)
+		{
+			return NIL;
+		}
+	}
+
+	/*
+	 * If the hash value we're looking for is known, we can search for the
+	 * interval directly.  That's fast and should only ever be the case for a
+	 * hash-partitioned table.
+	 */
+	if (prune->hashedEqualConsts)
+	{
+		int shardIndex = INVALID_SHARD_INDEX;
+		int shardCount = cacheEntry->shardIntervalArrayLength;
+		ShardInterval **sortedShardIntervalArray = cacheEntry->sortedShardIntervalArray;
+		char partitionMethod = context->partitionMethod;
+		FmgrInfo *compareFunction = cacheEntry->shardIntervalCompareFunction;
+		bool hasUniformHashDistribution = cacheEntry->hasUniformHashDistribution;
+
+		Assert(context->partitionMethod == DISTRIBUTE_BY_HASH);
+
+		shardIndex = FindShardIntervalIndex(prune->hashedEqualConsts->constvalue,
+											sortedShardIntervalArray,
+											shardCount, partitionMethod,
+											compareFunction, hasUniformHashDistribution);
+
+		if (shardIndex == INVALID_SHARD_INDEX)
+		{
+			return NULL;
+		}
+		else if (shardInterval &&
+				 sortedShardIntervalArray[shardIndex]->shardId != shardInterval->shardId)
+		{
+			return NULL;
+		}
+		else
+		{
+			return list_make1(sortedShardIntervalArray[shardIndex]);
+		}
+	}
+
+	/*
+	 * FIXME: Should we perform ExhaustivePrune() on the individual
+	 * shardInterval, in case there's both an = and a < / >? It'd probably be
+	 * better to pre-compute that by checking a PruningInstance's
+	 * less[Equal]Consts/greater[Equal]Consts against equalConsts.
+	 */
+	if (shardInterval)
+	{
+		return list_make1(shardInterval);
+	}
+
+	/*
+	 * For hashing, there's no further pruning methods available.  Because
+	 * triangle inequality doesn't hold for hashed values, we can't prune
+	 * based on < and >.
+	 */
+	if (context->partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		return NIL;
```

Comment:
Shouldn't this return all shards?

It seems like we could only get here if we have overlapping shards in a hash-partitioned table, might be worth eliminating that case in `PruneShardList` and adding an assert here.